eth2/types/src/beacon_state.rs

```diff
@@ -910,6 +883,79 @@ impl<T: EthSpec> BeaconState<T> {
         self.tree_hash_cache = BeaconTreeHashCache::default();
     }
 
+    /// Updates the cache if the provided epoch does not match the one in the cache.
+    pub fn update_proposer_indices_cache(
+        &mut self,
+        slot: Slot,
+        spec: &ChainSpec,
+    ) -> Result<(), Error> {
+        let slots_per_epoch = T::slots_per_epoch();
+        let epoch = slot.epoch(slots_per_epoch);
+
+        // Checking to see if epoch is something
+        if let Some(cached_epoch) = self.proposer_indices_cache.epoch {
+            // If the cached_epoch is the same as the one in the argument, then we don't need to update.
+            if cached_epoch == epoch {
+                return Ok(());
+            }
+        }
+
+        let indices = self.get_active_validator_indices(epoch);
+
+        if indices.is_empty() {
+            return Err(Error::InsufficientValidators);
+        }
+
+        // Using start_slot here just to make sure we are starting from the first slot in the epoch.
+        let start_slot = epoch.start_slot(slots_per_epoch);
+        let mut vector = vec![0; slots_per_epoch as usize];
+
+        // Iterating until we reach slots_per_epoch, caching the proposer index in the vector.
+        for slot_index in 0..slots_per_epoch as usize {
+            let seed = self.get_beacon_proposer_seed(start_slot + slot_index as u64, spec)?;
+            let mut i = 0;
+
+            // This was previously in the compute_proposer_index
+            loop {
+                let candidate_index = indices[compute_shuffled_index(
+                    i % indices.len(),
+                    indices.len(),
+                    &seed,
+                    spec.shuffle_round_count,
+                )
+                .ok_or(Error::UnableToShuffle)?];
+                let random_byte = {
+                    let mut preimage = seed.to_vec();
+                    preimage.append(&mut int_to_bytes8((i / 32) as u64));
+                    let hash = hash(&preimage);
+                    hash[i % 32]
+                };
+                let effective_balance = self.validators[candidate_index].effective_balance;
+                if effective_balance * MAX_RANDOM_BYTE
+                    >= spec.max_effective_balance * u64::from(random_byte)
+                {
+                    let elem = vector
+                        .get_mut(slot_index)
+                        .ok_or_else(|| Error::ProposerIndicesCacheIncomplete)?;
+                    *elem = candidate_index;
+                    break;
+                }
+                i += 1;
+            }
+        }
+
+        self.proposer_indices_cache = ProposerIndicesCache::new(
+            epoch,
+            FixedVector::new(vector).expect("should be of same length"),
+        );
+        Ok(())
+    }
```

Comment:
@michaelsproul This is my cache implementation. However, some tests are still failing (for example the `sanity_blocks` in the ef_tests).
I was wondering if you could take a loot at this implementation and tell me if it's correct. Atm I don't know if my implementation is incorrect, or if there's something going wrong with cache updating during the tests. Thanks! :D

Reply:
So I went through an intense debugging session and the main problem was just that I wasn't updating the cache!